### ISSUE TITLE
Szenario-Text bei Direct-Variante (einer CD-Aufgabe)

### ISSUE THREAD

**Issue by jvoigtlaender** [fmidue/modelling-tasks#206]
Im Sinne, dass man wenn man schon mit einem Klassendiagramm mit "sprechenden Namen" statt `A`, `B`, `C` ... arbeitet, man optional auch die Möglichkeit haben könnte, den eigentlichen Aufgabentext durch etwas Spezifischeres und auf eine Anwendungsdomäne Angepasstes ersetzen könnte.

Sinnvollster Einsatzzweck wäre im Moment wohl erstmal für eine Aufgabe aus der RepairCD/SelectValidCD-Familie, gegebenenfalls dem neuen Typ aus #205.

Es ergeben sich vielleicht noch verschiedene mögliche Design-Entscheidungen. Zunächst mal könnte man natürlich das Einfachste machen, also tatsächlich den Direct-Typ so nehmen wie er als Instanz aus dem Quiz-Typ entstünde. Also zum Beispiel wirklich den Aufgabentyp so interpretieren, dass es darum geht: "Hier ist ein falsches Klassendiagramm. Gib syntaktische Fehler/Probleme an." Da man allerdings volle Kontrolle über das Klassendiagramm hat, könnte man natürlich als Dozent sogar auch "semantische Fehler" einbauen und nach diesen fragen. Dafür müsste man natürlich diese geeignet in die Instanz-Definition einbringen können. (Besteht also erstmal die Frage, ob alle dafür denkbaren Fehler in den aktuellen "Datentyp für Fehlerbeschreibung" passen.) Auch denkbar wäre, dass man als Dozent noch die Möglichkeit hat, eigene Distraktoren (als Freitext) in die Liste der Antwortoptionen aufzunehmen.

Zunächst mal sollte sicher die einfachste Umsetzung gewählt werden, also wodurch am wenigsten Implementierungsaufwand entsteht.

**Comment by marcellussiegburg** [comment]
Eine einfache Version dieser Aufgabenvariante ist für den Aufgabentyp, der mit #205 eingeführt wurde, erstellt. Konfigurierbar ist der gesamte Aufgabentext (bis auf Higlighting-Info und Info zu vereinfacht dargestellten Klassen). Die [default-Konfigruation](https://autotool.fmi.uni-due.de/route/aufgabe/203#konfiguration) enthält den ursprünglichen Aufgabentext. Der Aufgabentext muss das Klassendiagram, die Liste der Beziehungen und die Gründe enthalten.

**Comment by jvoigtlaender** [comment]
Die Liste der Beziehungen muss nicht alle Beziehungen aus dem Klassendiagramm umfassen, richtig? Also man könnte ein Klassendiagramm vorgeben, das 10 Beziehungen enthält, aber für die Aufgabenstellung will man nur 4 davon benennen, von denen die Studierenden dann 2 identifizieren müssen, die zu einem bestimmten Problem führen.

Dann sollte das Feld `allRelationships` vielleicht lieber `relevantRelationships` heißen.

**Comment by jvoigtlaender** [comment]
In der Liste
```haskell
[ ( 1
        , ( True
          , Composition 
              { compositionName = "z"
              , compositionPart = LimitedLinking 
                  { linking = "D", limits = ( 1, Nothing ) }
              , compositionWhole = LimitedLinking 
                  { linking = "A", limits = ( 1, Just 1 ) } 
              } 
          ) 
        )
      , ( 2
        , ( True
          , Composition 
              { compositionName = "w"
              , compositionPart = LimitedLinking 
                  { linking = "B", limits = ( 1, Nothing ) }
              , compositionWhole = LimitedLinking 
                  { linking = "D", limits = ( 1, Just 1 ) } 
              } 
          ) 
        )
```
steht nach dem `Bool` immer eine Beziehung, die tatsächlich im Klassendiagramm vorkommt, oder? Also bei den erzeugten Instanzen ist es natürlich so, und ich kann mir auch nicht vorstellen, wie man den Aufgabentyp händisch nutzen würde, ohne diese Invariante einzuhalten (denn die Fragestellung ergibt dann ja wahrscheinlich gar keinen Sinn). Also da ist viel Redundanz, vor allem aber Potential für unbeabsichtigte Fehler. Also als Aufgabensteller ändere ich was an einer Stelle, aber nicht an anderer Stelle. Und plötzlich ist die Aufgabe "unlösbar".

Wäre es besser, wenn nach dem `Bool` stattdessen ein Index in die Liste `relationships` des `classDiagram`s kommt?

**Comment by jvoigtlaender** [comment]
Wenn ich in `errorReasons` als Keys statt `'a'`, `'b'`, `'c'` die Characters `'1'`, `'2'`, `'3'` hinschreibe, funktioniert die Lösungseingabe dann noch? (Also muss man da aufpassen? Wird die Eingabe fix auf "sind-Buchstaben, keine-Zahlen" geparsed, oder einfach genau die konkreten Characters akzeptiert, die in der `errorReasons`-Liste als Keys auftreten?)

**Comment by jvoigtlaender** [comment]
Wird bei der Angabe einer Direkt-Konfiguration irgendwie überprüft, dass ich in die `errorReasons`-Liste genau ein `True` und den Rest alles `False` habe? Falls nicht, was macht die Bewertungsfunktion dann daraus, da sie ja wahrscheinlich nur einen Singleton-Wert beim `reason` einlesen kann?

(Vorstellbar wäre durchaus, dass man die Aufgabe mit mehreren "possible reasons" stellen möchte. Also: "Hier ist ein Diagramm. Es enthält mehrere Fehler. Hier ist eine Liste von denkbaren Fehlern. Geben Sie einen tatsächlichen Fehler an. [Aber ich habe unter den zehn Antwortmöglichkeiten drei echte Fehler und sieben Distraktoren gemischt. Ich bin happy, wenn die Studierenden einen der drei echten Fehler finde. Welchen, ist mir egal.]" Würde das mit der Bewertungsfunktion funktionieren?)

**Comment by jvoigtlaender** [comment]
Wobei, bezogen auf den letzten Kommentar, natürlich ein Problem bestünde. Wenn sowohl `a` als auch `b` echte Fehler sind, sind dafür ja nicht unbedingt auf die gleichen Beziehungen schuldig. Aber in der Instanz kann ich ja gar nicht angeben, dass `a` wegen `1` und `2` und `b` wegen `3` und `4` vorliegt.

**Comment by jvoigtlaender** [comment]
Man könnte natürlich auch dem Aufgabensteller auferlegen, dass er nur dann verschiedene echte Fehlereinträge aufnehmen darf, wenn diese auf den gleichen Beziehungen beruhen. (Zum Beispiel könnte der Aufgabensteller dann für eigentlich den gleichen Fehler - mit den gleichen ursächlichen Beziehungen - zwei verschiedene Beschreibungen in die `reasons`-Liste aufnehmen, nicht wissend welche von beiden die Studierenden besser verstehen werden, aber beabsichtigend, dass beide als korrekte Antwort gewertet werden sollen.)

**Comment by jvoigtlaender** [comment]
Ein prinzipieller Vorschlag für eine konkrete Instanz dieses Aufgabentyps, eventuell (abstrahiert/vereinfacht/gekürzt/oberflächlich angerissen) für den Sammelband beim Punkt
> - nächster Schritt, wohl an Strategie "Szenario-Fragen" oder "Transferfragen" andockend, Einführung/Darstellung/Beschreibung/Diskussion des Direct-Aufgabentyps gemäß #206

des vorgeschlagenen Narrativs anführbar:

Vorgegebener Szenariotext aus Klausuraufgabe: 
![image](https://github.com/jvoigtlaender/modelling-tasks-internal/assets/5853832/51b1e6b3-bc69-49ab-b8a5-245073f5d8bb)

Vorgabe ebenso folgenden Klassendiagramms: 
![image](https://github.com/jvoigtlaender/modelling-tasks-internal/assets/5853832/accfd982-cf35-40db-b2ff-e4c1a692e678)

allerdings mit Multiplizät `1` rechts unten beim Professor.

Frage (natürlich ohne Hinweis auf die spezielle Multiplizität beim Professor): "Welches syntaktische oder semantische Problem besteht hier?"

Diverse Antwortmöglichkeiten, inklusive Distraktoren wie "Syntaxfehler [die aber gar nicht vorliegen]" und "Es gibt keine Instanzen [oder: Es existiert kein passendes Objektdiagramm]" etc., und eine tatsächlich korrekte Antwort im Sinne von "Es wurde eine Vorgabe des Szenarios nicht eingehalten [nämlich, dass nicht jeder Raum eine Prof-Büro ist; aber das muss nicht dastehen, es kann allgemeiner gehalten sein, und die Studierenden müssen selbst aktiv auf die hasOfficeIn-Beziehung zeigen]."

**Comment by jvoigtlaender** [comment]
Es ginge natürlich auch, dieses Klassendiagramm zu zeigen: 
![image](https://github.com/jvoigtlaender/modelling-tasks-internal/assets/5853832/b4d3a44f-6596-4b96-bfc3-1b091ec2b8e6)
wieder mit einer (anderen) Änderung der Multiplizitäten an der untersten Beziehung, und dann ist auch wieder herauszufinden, dass das das Problem ist (weil plötzlich ein Prof ohne Büro dastehen könnte; und als ein Distraktor taucht in der Liste vielleicht die umgekehrte Behauptung auf, dass manche keinen darin sitzenden Professor hätten o.ä.).

**Comment by jvoigtlaender** [comment]
Oder man zeigt eines der mit korrekten Multiplizitäten versehenen Klassendiagramme, aber ändert darin eine Vererbung zu Assoziation oder umgekehrt (also im Grunde einfach nur "falsche Pfeilspitze an einer Stelle"), und lässt die Studierenden herausfinden, welcher Teil des zu modellierenden Szenarios dadurch "kaputtgeht" (ohne dass sie vorab wissen, dass und welche Pfeilspitze falsch ist, sondern darauf müssen sie erstmal kommen, indem sie das Klassendiagramm mit dem Szenariotext vergleichen).

**Comment by jvoigtlaender** [comment]
Und noch eine anders gelagerte "Szenario-Situation", die sich zu einer konkreten Instanz hier heranziehen ließe:

Man gebe in etwa dieses Klassendiagramm aus unserem Übungsbetrieb vor (vielleicht für Präsentationszwecke wieder etwas eingedampft, obwohl man in Wirklichkeit wahrscheinlich gerade komplexere Diagramme einsetzen würde, in denen die Studierenden erstmal den problematischen Ausschnitt identifizieren müssen): 
![image](https://github.com/jvoigtlaender/modelling-tasks-internal/assets/5853832/5934c6c9-69fa-4808-a2f9-621778af76c0)

und sage: "Es soll ein Objektdiagramm hierfür gebildet werden. Es ergibt sich so, dass gerade keine Säle vorhanden sind, weil das Theater im Umbau ist. Aber gebuchte Termine für Veranstaltungen möge es bitte schon geben. Welches Problem ergibt sich?" Unter den potentiellen Antwortmöglichkeiten könnte sein, dass die Komposition zwischen Theater und Saal ein Problem darstelle (stimmt aber nicht, da ja implizit nur am Theater-Ende eine 1 steht, am Saal-Ende aber eine 0..*). Also das ist ein Distraktor. 

Aber andere Fehlermöglichkeiten existieren tatsächlich. Eine davon ist, dass zu einer Buchung ein Termin gehören muss, und zu einem Termin ein Saal. Bähm.

Oder dass eine Buchung mindestens einen Platz haben muss, und ein Platz in einem Saal sein muss. Bähm. (Oder man packt an diese Komposition eine 0..1, damit nur der andere echte Fehler existiert, und dies hier zum Distraktor wird.)

**Comment by jvoigtlaender** [comment]
Eventuell lässt sich der Aufgabentyp auch allgemeiner nutzen, um andere "Analyse-Fragen" geeignet einzukleiden.

Etwa zu dem Theater-Beispiel Fragen wie:
- Gibt es dazu Objektdiagramme mit Buchungen aber ohne Säle? (Wenn nein, warum nicht?)
- Gibt es dazu Objektdiagramme mit Sälen aber ohne Buchungen? (Wenn nein, warum nicht?)
- ... ähnliche Fragen für Buchungen, Termine, Veranstaltungen [oder subtiler, Filme statt Veranstaltungen] ...

**Comment by jvoigtlaender** [comment]
Mit etwas Abstand könnte man sich wahrscheinlich fragen, was eigentlich der Gewinn dieses Aufgabentyps ist, wenn dabei gar nichts generiert wird, wenn also das Gleiche auch in einem beliebigen Multiple-Choice-Aufgabentyp umsetzbar wäre (also etwa einen generischen Aufgabentyp in Moodle nehmen, dort ein Bild eines Klassendiagramms hochladen, Text hinzufügen, eine Auswahlliste einstellen).

Ich schätze, es gibt mindestens zwei Antworten:
1. Einbettungsmöglichkeit in das Umfeld (E-Learning-System), in dem auch die anderen, automatisch generierten Aufgaben gestellt werden.
2. Profitieren von unseren bereits vorhandenen Möglichkeiten der "Verschleierung" bzw. low-level-Variierung. Zum Beispiel könnten hier ja aus einer einzigen Konfiguration von `classDiagram`, `allRelationships` und `errorReasons` zumindest oberflächlich recht verschiedene Aufgabeninstanzen angezeigt werden, indem das `IncorrectCd` layoutmäßig perturbiert wird und die `RelationshipsList` sowie `ReasonsList` jeweils auch zufallsbasiert permutiert; alles mit automatisch sich anpassender Bewertung.

**Comment by marcellussiegburg** [comment]
> Also da ist viel Redundanz, vor allem aber Potential für unbeabsichtigte Fehler. Also als Aufgabensteller ändere ich was an einer Stelle, aber nicht an anderer Stelle. Und plötzlich ist die Aufgabe "unlösbar".

Ja, das stimmt, hier sind viele Redundanzen. Das Fehlerpotenzial wird durch die Checker gemindert, die es verbieten, dass eine der Beziehungen an einer stelle aber nicht an der anderen auftauchen.

> Wäre es besser, wenn nach dem Bool stattdessen ein Index in die Liste relationships des classDiagrams kommt?

Ja, das wäre auch denkbar. Der Nachteil wäre, dass es für die Aufgabenerstellung weniger übersichtlich ist, welche Beziehung gemeint ist. Das ist sicherlich eine Abwägungsentscheidung.

**Comment by jvoigtlaender** [comment]
Okay, wenn die Checker das abfangen, ist Copy-and-Paste wohl nicht so gefährlich.

**Comment by marcellussiegburg** [comment]
> Die Liste der Beziehungen muss nicht alle Beziehungen aus dem Klassendiagramm umfassen, richtig? Also man könnte ein Klassendiagramm vorgeben, das 10 Beziehungen enthält, aber für die Aufgabenstellung will man nur 4 davon benennen, von denen die Studierenden dann 2 identifizieren müssen, die zu einem bestimmten Problem führen.

Bisher ist es so, dass dort alle auftauchen müssen. Wenn das geändert würde, wäre sicherlich auch diese Umbenennung sinnvoll:

> Dann sollte das Feld `allRelationships` vielleicht lieber `relevantRelationships` heißen.

Ich deute daraus, dass der Wunsch wäre, partielle Listen angeben zu können? Wenn das der Fall ist, wäre der vorgeschlagene Wechsel zur Indexiertenvariante vielleicht besser...

> Wäre es besser, wenn nach dem Bool stattdessen ein Index in die Liste relationships des classDiagrams kommt?

**Comment by jvoigtlaender** [comment]
> Ich deute daraus, dass der Wunsch wäre, partielle Listen angeben zu können?

Ich denke schon. Denn bei einer Szenario-Aufgabe möchte man als Aufgabensteller wahrscheinlich die Möglichkeit haben, nicht unbedingt alle Beziehungen aufzählen zu müssen. (Zum Beispiel hat man vielleicht eine inhaltliche/semantische Fragestellung, die sich nur auf die Vererbungsstruktur im Modell bezieht, aber das Modell enthält auch noch Assoziationen etc.)

**Comment by marcellussiegburg** [comment]
> Wird bei der Angabe einer Direkt-Konfiguration irgendwie überprüft, dass ich in die errorReasons-Liste genau ein True und den Rest alles False habe? Falls nicht, was macht die Bewertungsfunktion dann daraus, da sie ja wahrscheinlich nur einen Singleton-Wert beim reason einlesen kann?

Da fehlt im Moment tatsächlich noch ein check. Die Bewertung wäre nicht sinnvoll.

> [Aber ich habe unter den zehn Antwortmöglichkeiten drei echte Fehler und sieben Distraktoren gemischt. Ich bin happy, wenn die Studierenden einen der drei echten Fehler finde. Welchen, ist mir egal.]" Würde das mit der Bewertungsfunktion funktionieren?

Das würde im Moment nicht funktionieren. Die Bewertungsfunktion müsste dafür explizit geschrieben werden. Und wahrscheinlich sollte die Aufgabenkonfiguration dann explizit `AnyOf` (oder ähnliches) an die Liste annotieren, damit bei der Aufgabenerstellung auch klar ist, was das bedeuten soll.

**Comment by marcellussiegburg** [comment]
> Wenn ich in errorReasons als Keys statt 'a', 'b', 'c' die Characters '1', '2', '3' hinschreibe, funktioniert die Lösungseingabe dann noch? (Also muss man da aufpassen? Wird die Eingabe fix auf "sind-Buchstaben, keine-Zahlen" geparsed, oder einfach genau die konkreten Characters akzeptiert, die in der errorReasons-Liste als Keys auftreten?)

Die Antwort ist wohl, es funktioniert dann nicht mehr so intuitiv. Das hängt damit zusammen, wie yaml-geparst wird bzw. getypt ist.
Die Eingabe `reason: 1` würde abgelehnt werden mit einer Meldung, dass der Typ nicht stimmt, stattdessen müsste dann `reason: '1'` eingegeben werden. Daher ist hier wohl Vorsicht geboten (und der Checker sollte nur `[a-zA-Z]` zulassen?)

**Comment by jvoigtlaender** [comment]
Ja, da sollte wohl der Checker aktiv werden, um Fehleinstellungen schon in der Konfiguration zu vermeiden.
